core/src/WsTrust/MexParser.cs

```diff
@@ -0,0 +1,280 @@
+// furnished to do so, subject to the following conditions :
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+//------------------------------------------------------------------------------
+
+using System;
+using System.Collections.Generic;
+using System.Globalization;
+using System.Linq;
+using System.Threading.Tasks;
+using System.Xml.Linq;
+using Microsoft.Identity.Core.Http;
+
+namespace Microsoft.Identity.Core.WsTrust
+{
+    internal enum WsTrustVersion
+    {
+        WsTrust13,
+        WsTrust2005
+    }
+
+    internal class WsTrustAddress
+    {
+        public Uri Uri { get; set; }
+
+        public WsTrustVersion Version { get; set; }
+    }
+
+    internal class MexPolicy
+    {
+        public WsTrustVersion Version { get; set; }
+
+        public string Id { get; set; }
+
+        public UserAuthType AuthType { get; set; }
+
+        public Uri Url { get; set; }
+    }
+
+    internal static class MexParser
+    {
+        private const string WsTrustSoapTransport = "http://schemas.xmlsoap.org/soap/http";
+
+        public static async Task<WsTrustAddress> FetchWsTrustAddressFromMexAsync(string federationMetadataUrl, UserAuthType userAuthType, RequestContext requestContext)
+        // This method would return null if the wanted policy is not found in mexDocument. It could potentially throw XmlException or http-relevant exceptions too.
+        {
+            XDocument mexDocument = await FetchMexAsync(federationMetadataUrl, requestContext).ConfigureAwait(false);
+            return ExtractWsTrustAddressFromMex(mexDocument, userAuthType, requestContext);
+        }
+
+        internal static async Task<XDocument> FetchMexAsync(string federationMetadataUrl, RequestContext requestContext)
+        {
+            var httpResponse = await HttpRequest.SendGet( // It would potentially throw MsalServiceException
+                new UriBuilder(federationMetadataUrl).Uri, null, requestContext).ConfigureAwait(false);
+            if (httpResponse.StatusCode != System.Net.HttpStatusCode.OK) {
+                throw new Client.MsalException(
+                    MsalError.AccessingWsMetadataExchangeFailed,
+                    string.Format(CultureInfo.CurrentCulture,
+                        "Response status code does not indicate success: {0} ({1}).",
+                        (int)httpResponse.StatusCode, httpResponse.StatusCode));
```

Comment:
>(int)httpResponse.StatusCode, httpResponse.StatusCode [](start = 24, length = 53)

? what benefit does this do?

Reply:
You mean why seemingly duplicates 2 `httpResponse.StatusCode` here? It displays both an integer http status code and its descriptive name. You can see [its effect in this pre-existing test case](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1118/files#diff-89254dc64c4c22af332e70493e58486eR250), which is also the reason that we need to throw this exception, with this error message, here.

Reply:
Copied from earlier comment in github:

You mean why seemingly duplicates 2 `httpResponse.StatusCode` here? It displays both an integer http status code and its descriptive name. You can see [its effect in this pre-existing test case](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1118/files#diff-89254dc64c4c22af332e70493e58486eR250), which is also the reason that we need to throw this exception, with this error message, here.

---
In reply to: [201755264](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1118#discussion_r201755264) [](ancestors = 201755264)